orchagent/neighorch.cpp

```diff
@@ -57,13 +59,102 @@ bool NeighOrch::addNextHop(IpAddress ipAddress, string alias)
     NextHopEntry next_hop_entry;
     next_hop_entry.next_hop_id = next_hop_id;
     next_hop_entry.ref_count = 0;
+    next_hop_entry.nh_flags = 0;
+    next_hop_entry.if_alias = alias;
     m_syncdNextHops[ipAddress] = next_hop_entry;
 
     m_intfsOrch->increaseRouterIntfsRefCount(alias);
 
     return true;
 }
 
+bool
+NeighOrch::setNextHopFlag (const IpAddress &ipaddr, const uint32_t nh_flag)
+{
+    SWSS_LOG_ENTER();
+
+    bool rc = false;
+
+    assert(hasNextHop(ipaddr));
+
+    if (m_syncdNextHops[ipaddr].nh_flags & nh_flag) {
+        return (true);
+    }
+
+    m_syncdNextHops[ipaddr].nh_flags |= nh_flag;
+
+    switch (nh_flag) {
+        case NHFLAGS_IFDOWN:
+            rc = route_orch->invalidnexthopinNextHopGroup(ipaddr);
+            break;
+        default:
+            assert(0);
+            break;
+    }
+
+    return (rc);
+}
+
+bool
+NeighOrch::clearNextHopFlag (const IpAddress &ipaddr, const uint32_t nh_flag)
+{
+    SWSS_LOG_ENTER();
+
+    bool rc = false;
+
+    assert(hasNextHop(ipaddr));
+
+    if (!(m_syncdNextHops[ipaddr].nh_flags & nh_flag)) {
+        return (true);
+    }
+
+    m_syncdNextHops[ipaddr].nh_flags &= ~nh_flag;
+
+    switch (nh_flag) {
+        case NHFLAGS_IFDOWN:
+            rc = route_orch->validnexthopinNextHopGroup(ipaddr);
+            break;
+        default:
+            assert(0);
+            break;
+    }
+
+    return (rc);
+}
+
+bool
+NeighOrch::isNextHopFlagSet (const IpAddress &ipaddr, const uint32_t nh_flag)
+{
+    SWSS_LOG_ENTER();
+
+    assert(hasNextHop(ipaddr));
+    if (m_syncdNextHops[ipaddr].nh_flags & nh_flag) {
+        return (true);
+    }
+
+    return (false);
+}
+
+bool
+NeighOrch::ifChangeInformNextHop (const string &alias, bool if_up)
+{
+    SWSS_LOG_ENTER();
+
+    for (auto nhop = m_syncdNextHops.begin(); nhop != m_syncdNextHops.end(); ++nhop) {
```

Comment:
How big do you expect the nexthop set to be? Typical data center won't have more than 100.

Reply:
Ok. Since it is handled in the same context as link down event, my concern was if it could result in some delays/side effects. The other option would have been acting on the link down notification from kernel. Closing this as the assumption is, size of nexthop set would be comparatively smaller.